mcstatus/pinger.py

```diff
@@ -149,125 +153,4 @@ async def test_ping(self) -> float:
-            else:
-                self.sample = None
-
-    class Version:
-        name: str
-        protocol: int
-
-        def __init__(self, raw: dict[str, object]):
-            if not isinstance(raw, dict):
-                raise ValueError(f"Invalid version object (expected dict, found {type(raw)})")
-
-            if "name" not in raw:
-                raise ValueError("Invalid version object (no 'name' value)")
-            if not isinstance(raw["name"], str):
-                raise ValueError(f"Invalid version object (expected 'name' to be str, was {type(raw['name'])})")
-            self.name = raw["name"]
-
-            if "protocol" not in raw:
-                raise ValueError("Invalid version object (no 'protocol' value)")
-            if not isinstance(raw["protocol"], int):
-                raise ValueError(f"Invalid version object (expected 'protocol' to be int, was {type(raw['protocol'])})")
-            self.protocol = raw["protocol"]
-
-    players: Players
-    version: Version
-    description: str
-    favicon: Optional[str]
-    latency: float = 0
-
-    def __init__(self, raw: dict[str, object]):
-        self.raw = raw
-
-        # TODO: Consider using a TypedDict here, to avoid all of the type casts
-        if "players" not in raw:
-            raise ValueError("Invalid status object (no 'players' value)")
-        self.players = PingResponse.Players(cast("dict[str, object]", raw["players"]))
-
-        if "version" not in raw:
-            raise ValueError("Invalid status object (no 'version' value)")
-        self.version = PingResponse.Version(cast("dict[str, object]", raw["version"]))
-
-        if "description" not in raw:
-            raise ValueError("Invalid status object (no 'description' value)")
-        self.description = self._parse_description(cast("dict[str, object]", raw["description"]))
-
-        self.favicon = cast(Optional[str], raw.get("favicon"))
-
-    @staticmethod
-    def _parse_description(raw_description: Union[dict, list, str]) -> str:
-        if isinstance(raw_description, str):
-            return raw_description
-
-        if isinstance(raw_description, dict):
-            entries = raw_description.get("extra", [])
-            end = raw_description["text"]
-        else:
-            entries = raw_description
-            end = ""
-
-        description = ""
-
-        for entry in entries:
-            for style_key, style_val in STYLE_MAP.items():
-                if entry.get(style_key):
-                    try:
-                        if isinstance(style_val, dict):
-                            style_val = style_val[entry[style_key]]
-
-                        description += f"§{style_val}"
-                    except KeyError:
-                        pass  # ignoring these key errors strips out html color codes
-            description += entry.get("text", "")
-
-        return description + end
+PingResponse: TypeAlias = JavaStatusResponse
```

Comment:
`PingResponse` used in type hints, and because we have `py.typed` file since #332, we should count type hints API as public API.

After deprecation period you can safely remove this type alias.

Reply:
I do agree that we should not break type-hints of people who like to specify them, and we should keep a deprecated type-alias for this.

Although what's a bit more problematic here is that using this type-alias wont' produce a deprecation warning, so people who update the library without checking the release notes won't necessarily know that `PingResponse` has became deprecated.

However I don't see a way we could produce this warning on use, as this is just a type alias, so just keeping it in like this is probably good enough.

Reply:
Thinking of this though, since we won't be able to produce a deprecation warning here, keeping `PingResponse` typealias here will mean that people can keep using it, and won't know that it's actually deprecated. So eventually, after some deprecation periods, this would become a breaking change anyway.

So the question is whether we want to keep this and remove it with other deprecations, even though it's gonna be essentially a breaking change unless people check the release notes, or whether we should just remove it as it won't have deprecation period anyway, and not many people are actually checking these release notes here. We could do a major version bump with this minor breaking change I suppose.

Reply:
> keeping `PingResponse` typealias here will mean that people can keep using it, and won't know that it's actually deprecated.

I don't agree. Yes - we can't raise warning here, but this is still deprecation, and it would be confusing to keep it here, even after removing the main deprecations.

> won't know that it's actually deprecated.

If they use type hints, they 100% use actual code, so they will know about deprecations from places where we can raise it.

> So eventually, after some deprecation periods, this would become a breaking change anyway.

I would better count even deprecations, as breaking changes because it's just a workaround, so old code will work.

> So the question is whether we want to keep this and remove it with other deprecations, even though it's gonna be essentially a breaking change unless people check the release notes, or whether we should just remove it as it won't have deprecation period anyway, and not many people are actually checking these release notes here.

I think, that all what we can do here - document deprecation inside docstring.

Reply:
Alright, that's fair

Reply:
It is enough to warrant keeping it. For example
```py
def get_server_status(ip: str) -> PingResponse:
    return mcstatus.JavaServer.lookup(ip).status()
```

Reply:
Where did you see that in this branch?

Reply:
It's an example user code, not ours.

Reply:
@kevinkjt2000, It's an example user code, not ours.

Reply:
In lieu of a warning, an ImportError and/or type checker error would force them to swap to the new type. Otherwise, they'll just be surprised later anyway if we never provide a warning. I see a few ways this could go down:

1. We surprise users later
2. We surprise them now
3. We give them a warning
4. Major API version bump

Obviously, 1 and 2 are terrible choices and I was advocating for 2 or 4 earlier. 3 should be possible if we do a deep copy of the class and call our deprecated wrapper for the `__init__` method. Now, I'm in camp 3.

Reply:
The issue is that a warning that's raised based on a signature won't appear to the users who're importing this just for type-hinting (which is likely the majority, most people aren't making their own instances of the ping response class).

The conversation I had with @PerchunPak above was actually precisely about this, that you can't deprecate a type annotation, and so this will inevitably become a breaking change, so we decided that it should stay for now, and perhaps get removed once we have a major version bump.

Whether this should be kept in with full deprecation of `__init__` is a different question all together, I'd argue that it probably doesn't need to, since I wouldn't necessarily consider the initialization of `PingResponse` to be something the users should be doing, therefore it shouldn't be a part of our public API which needs deprecation (or should it be?).

Reply:
I agree with ItsDrinke, but just wanted that 1. and 4. are the same...